### PR TITLE
Ojo

### LINKED ISSUE #587: Oho, Neutral Killing
Each night choose a role, everyone with that role is attacked (armor piercing)
You visit every person with that role

Trackers can catch you really easily

### PR DESCRIPTION
**Pull request by ItsSammyM**
closes #587

### CODE REVIEW

File: client/src/components/RoleDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+export default function RoleDropdown(props: {
+    value: Role,
+    onChange: (role: Role) => void,
+    disabledRoles?: Role[]
+}): ReactElement {
+
+
+    let options = Object.keys(ROLES)
+        .filter((role)=>!props.disabledRoles?.includes(role as Role))
```
Comment (Jack-Papel) [inline client/src/components/RoleDropdown.tsx:15]: After I merge buttons PR there will be a getAllRoles function in roleListState.d.tsx

File: client/src/components/RoleDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        .map((role)=>{
+        return <option value={role}>{translate("role."+role+".name")}</option>
+    })
```
Comment (Jack-Papel) [inline client/src/components/RoleDropdown.tsx:18]: epic indentation

File: client/src/components/RoleDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    
+    return <select
+        value={props.value}
+        onChange={(e)=>{
+            props.onChange(e.target.value as Role)
+        }}>
```
Comment (Jack-Papel) [inline client/src/components/RoleDropdown.tsx:24]: nit:
`onChange={e => props.onChange(e.target.value as Role)}`

File: client/src/components/RoleDropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import translate from "../game/lang"
+
+export default function RoleDropdown(props: {
+    value: Role,
+    onChange: (role: Role) => void,
+    disabledRoles?: Role[]
+}): ReactElement {
+
+
+    let options = Object.keys(ROLES)
```
Comment (Jack-Papel) [inline client/src/components/RoleDropdown.tsx:14]: Can be inlined

File: client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    switch(props.action.type){
+        case "none":
+            return <>{actionTypeDropdown}</>
+        case "kill":
+            return <>{actionTypeDropdown}{roleDropdown}</>
+        case "see":
+            return <>{actionTypeDropdown}{roleDropdown}</>
+    }
+}
```
Comment (Jack-Papel) [inline client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx:42]: This entire function can be inlined into the return statement

File: client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+function ActionTypeDropdown(props: {
+    action: OjoAction,
+    onChange: (action: OjoAction) => void
+}): ReactElement {
+    let options = [
```
Comment (Jack-Papel) [inline client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx:48]: Can be inlined, would be cleaner too.

File: client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                case "none":
+                    props.onChange({type: e.target.value as "kill" | "see", role: "amnesiac"})
+                    break;
+                case "see":
+                case "kill":
+                    props.onChange({type: e.target.value as "kill" | "see", role: props.action.role})
+                    break;
+            }
+
+
```
Comment (Jack-Papel) [inline client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx:71]: All your PRs have so much random whitespace...

File: client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx (hunk 5 above)
Comment (Jack-Papel) [reply client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx:42]: Also there is no text here. As a player I don't know why my player list has this thing on top of it.

File: client/src/resources/lang/en_us.json [hunk 8]
```diff
@@ -1035,6 +1044,11 @@
...
 
+    "wiki.article.role.ojo.guide":"- You may choose to either kill every player with a certain role, or find out every player with a certain role\n\n_For Example_\n- You choose to kill all lookouts and then learn that @1 and @2 are lookouts, and then they die\n- You choose to see all trackers, and you see nobody, so you know there is no tracker",
+    "wiki.article.role.ojo.abilities":"- You may choose to do one of the following\n    - See all players with a certain role, you visit all players with that role\n    - Kill all players with a certain role, you visit all players with that role with an attacking visit\n        - Players you visit with an attacking visit will be attacked with an armor-piercing attack",
+    "wiki.article.role.ojo.attributes":"- You are told the name and role of all players you visit, weather you choose see or kill\n- If you are dead, you can not choose to see or kill",
```
Comment (Jack-Papel) [inline client/src/resources/lang/en_us.json:1049]: *whether

File: client/src/resources/lang/en_us.json [hunk 9]
```diff
@@ -1035,6 +1044,11 @@
...
+    "wiki.article.role.ojo.abilities":"- You may choose to do one of the following\n    - See all players with a certain role, you visit all players with that role\n    - Kill all players with a certain role, you visit all players with that role with an attacking visit\n        - Players you visit with an attacking visit will be attacked with an armor-piercing attack",
+    "wiki.article.role.ojo.attributes":"- You are told the name and role of all players you visit, weather you choose see or kill\n- If you are dead, you can not choose to see or kill",
+    "wiki.article.role.ojo.extra":"- Your ability will work on the person you actually visited, so if a transporter transports a player you visit with someone else, you will see or attack the other player\n- A Witch can force you to visit someone, if that happens, you wont attack, because the witch doesn't force you to use an attacking visit\n- You don't visit yourself\n- You don't visit dead players",
```
Comment (Jack-Papel) [inline client/src/resources/lang/en_us.json:1050]: In this case, do you get their actual role, or does it tell you that they are the role you investigated?

File: server/src/game/role/ojo.rs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                            })
+                        } else {
+                            None
+                        }
+                    )
+                    .collect()
+                );
+            }
+            Priority::Kill => {
+                if matches!(self.chosen_action, OjoAction::Kill{..}) {
```
Comment (Jack-Papel) [inline server/src/game/role/ojo.rs:71]: nit: might as well use if...let

**Review by Jack-Papel — APPROVED** [review]
Approved, but PLEASE ADDRESS COMMENTS!

File: client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx (hunk 7 above)
Comment (ItsSammyM) [reply client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx:71]: Genuinely help me

File: client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx (hunk 5 above)
Comment (ItsSammyM) [reply client/src/menu/game/gameScreenContent/RoleSpecificMenus/SmallOjoMenu.tsx:42]: Can you help me with this. This also needs to be fixed for amnesiac.

**Pull request merged**